File: DealFinder.Web/src/shared/deals/deals.mapper.ts
import { GetDealsByLocationResponse } from './getDealsByLocationResponse';
import { Deal } from './deal';
import { DealRecord } from './deal.record';

export class DealsMapper {
    private static ONE_MILE_IN_METERS: number = 1609;

    static map(payload: GetDealsByLocationResponse): Deal[] {
        let response = [];

        if (payload.hasError)
            return response;

        payload.deals.forEach((item: DealRecord) => {
            response.push({
                id: item.id,
                title: item.title,
                summary: item.summary,
                distanceDescription: this.generateAppropriateDistanceUnit(item.distanceInMeters),
                createdAt: new Date(item.createdAt),
                location: {
                    latitude: item.location.latitude,
                    longitude: item.location.longitude
                },
                user: {
                    identifier: item.user.identifier,
                    picture: item.user.picture,
                    username: item.user.username
                },
                votes: {
                    totalVotes: item.votes.totalVotes,
                    positiveVotes: item.votes.positiveVotes,
                    negativeVotes: item.votes.negativeVotes,
                    finalScore: item.votes.finalScore,
                    hasAlreadyVoted: item.votes.hasAlreadyVoted
                }
            });
        });

        return response;
    }

    private static generateAppropriateDistanceUnit(distanceInMeters: number): string {
        let distanceInMiles = Math.floor(distanceInMeters / this.ONE_MILE_IN_METERS);

        if (distanceInMiles > 0)
            return `${distanceInMiles} mile${distanceInMiles === 1 ? '' : 's'} away`;

        return `${Math.floor(distanceInMeters)} meter${distanceInMeters === 1 ? '' : 's'} away`;
    }
}

File: DealFinder.Data/Deals/DealsMapper.cs
using System.Collections.Generic;
using DealFinder.Core.Distance;

namespace DealFinder.Data.Deals
{
    public class DealsMapper
    {
        public static List<DealModel> Map(List<DealRecord> dealRecords)
        {
            var response = new List<DealModel>();

            foreach (var dealRecord in dealRecords)
            {
                response.Add(new DealModel
                {
                    Title = dealRecord.Title,
                    Summary = dealRecord.Summary,
                    DistanceInMeters = dealRecord.DistanceInMeters,
                    Location = new Location
                    {
                        Longitude = dealRecord.Longitude,
                        Latitude = dealRecord.Latitude
                    }
                });   
            }

            return response;
        }
    }
}